### PR TITLE
Location Types

### PR DESCRIPTION
**Pull request by DerDrodt**
This PR is a re-implementation of the location type system currently used in ABS.

The principles and main types remain the same, while the inference is now simpler, faster, and rule-based.

### CODE REVIEW

**Review by rudi — CHANGES_REQUESTED** [review]
The directory `frontend/.idea` should probably not be part of the pull request

File: frontend/src/main/java/org/abs_models/Absc.java [hunk 1]
```diff
@@ -196,12 +193,12 @@ public LocationType convert(String value) throws Exception {
...
 
 
-    @Option(names = { "--locscope" },
+    /*@Option(names = { "--locscope" },
             description = "sets the location aliasing scope (allowed values: ${COMPLETION-CANDIDATES}) (default: ${DEFAULT-VALUE})",
             paramLabel = "scope")
     // default value taken from declaration of
     // LocationTypeInferrerExtension.precision
-    public LocationTypeInferrerExtension.LocationTypingPrecision locationTypeScope = LocationTypingPrecision.CLASS_LOCAL_FAR;
+    public LocationTypeInferrerExtension.LocationTypingPrecision locationTypeScope = LocationTypingPrecision.CLASS_LOCAL_FAR;*/
```
Comment (rudi) [inline frontend/src/main/java/org/abs_models/Absc.java:201]: If the option is unused, this block can be removed.

File: frontend/src/main/java/org/abs_models/frontend/typechecker/locationtypes/Benchmark.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
+package org.abs_models.frontend.typechecker.locationtypes;
```
Comment (rudi) [inline frontend/src/main/java/org/abs_models/frontend/typechecker/locationtypes/Benchmark.java:1]: This file looks a bit like it's not necessary after the merge?  We should remove the old location type system since it bit-rotted -- any dead code will lead to maintainer confusion down the road.

File: frontend/src/main/java/org/abs_models/frontend/typechecker/locationtypes/Constraints.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
+package org.abs_models.frontend.typechecker.locationtypes;
```
Comment (rudi) [inline frontend/src/main/java/org/abs_models/frontend/typechecker/locationtypes/Constraints.java:1]: This class maybe wants to be renamed?  There's `Constraint` and `Constraints` with only one letter difference, and no documentation in either.  In general, a file / class header that says something like "This class is part of the XXX system.  Its purpose is XXX, which it does together with YYY.  The main entry points are ZZZ." would be appreciated.

File: frontend/src/main/java/org/abs_models/frontend/typechecker/locationtypes/LocationTypeExtension.java [hunk 4]
```diff
@@ -1,157 +1,139 @@
-/** 
- * Copyright (c) 2009-2011, The HATS Consortium. All rights reserved. 
+/**
```
Comment (rudi) [inline frontend/src/main/java/org/abs_models/frontend/typechecker/locationtypes/LocationTypeExtension.java:1]: Can/should this old implementation be removed?

File: frontend/src/test/java/org/abs_models/frontend/typesystem/LocationTypeTests.java [hunk 5]
```diff
@@ -228,35 +227,22 @@ public void callReturn() {
...
-                      "module M.S2; import * from M.S1; { I i; i = new C(); i!m(); }", LocationType.FAR);
+            "module M.S2; import * from M.S1; { I i; i = new C(); i!m(); }", LocationType.FAR);
     }
-    
+
     @Test
     public void typeSyn() {
         assertInferOk("interface I {} type I2 = [Somewhere] I; { I2 i; i = null; }", LocationType.SOMEWHERE);
     }
-
```
Comment (rudi) [inline frontend/src/test/java/org/abs_models/frontend/typesystem/LocationTypeTests.java:238]: Can you comment on why the test was removed?  (Inapplicable, fails with new implementation, ...)

File: frontend/src/test/java/org/abs_models/frontend/typesystem/LocationTypeTests.java [hunk 6]
```diff
@@ -318,38 +304,11 @@ public void testInferenceRetypeChecking() {
         SemanticConditionList e = m.typeCheck();
         m.typeCheck(new SemanticConditionList());
     }
-
```
Comment (rudi) [inline frontend/src/test/java/org/abs_models/frontend/typesystem/LocationTypeTests.java:321]: Can you comment on why the test was removed?  (Inapplicable, fails with new implementation, ...)

File: frontend/src/test/java/org/abs_models/frontend/typesystem/LocationTypeTests.java (hunk 5 above)
Comment (DerDrodt) [reply frontend/src/test/java/org/abs_models/frontend/typesystem/LocationTypeTests.java:238]: These two tests are based on the parameterized far types, which were removed. Without these far types, they fail.

File: frontend/src/test/java/org/abs_models/frontend/typesystem/LocationTypeTests.java (hunk 6 above)
Comment (DerDrodt) [reply frontend/src/test/java/org/abs_models/frontend/typesystem/LocationTypeTests.java:321]: I was unsure what exactly was tested here. There are no assertions, it always failed due to the default type being set to INFER, which led to errors, and I believe it also failed for the old system.

File: frontend/src/main/java/org/abs_models/frontend/typechecker/locationtypes/Benchmark.java (hunk 2 above)
Comment (DerDrodt) [reply frontend/src/main/java/org/abs_models/frontend/typechecker/locationtypes/Benchmark.java]: Yes, that class was just to compare the systems. It is now removed.

File: frontend/src/main/java/org/abs_models/frontend/typechecker/locationtypes/LocationTypeExtension.java (hunk 4 above)
Comment (DerDrodt) [reply frontend/src/main/java/org/abs_models/frontend/typechecker/locationtypes/LocationTypeExtension.java:1]: The entire file? Possibly. It is currently used. After inference, the entire model is checked again by the LocationTypeExtension, which ensures that all inferred types are correct. It also is used when inference is disabled.

Maybe the inference system can cover all checks, but I think it is better to check twice and to leave inference optional.

File: frontend/src/main/java/org/abs_models/frontend/typechecker/locationtypes/Constraints.java (hunk 3 above)
Comment (DerDrodt) [reply frontend/src/main/java/org/abs_models/frontend/typechecker/locationtypes/Constraints.java:1]: That probably is confusing, and an explanation makes a lot of sense. I will do that.

Comment (rudi) [reply frontend/src/main/java/org/abs_models/frontend/typechecker/locationtypes/Constraints.java:1]: Many thanks! In general, some short bits of documentation for added classes and aspects help a lot.  Didn't bother to write this at all points, just where it jumped out at me most.

File: frontend/src/main/java/org/abs_models/frontend/typechecker/locationtypes/LocationTypeExtension.java (hunk 4 above)
Comment (rudi) [reply frontend/src/main/java/org/abs_models/frontend/typechecker/locationtypes/LocationTypeExtension.java:1]: Ok, got it.  That said, if location type inference mutates the AST, maybe it should not be a type checker extension at all?  We have the horrible `Main.analyzeFlattenAndRewriteModel` method that mangles the AST in all kinds of ways; any mutation of the AST should be in there.

(The `Main.analyzeFlattenAndRewriteModel` method is the product of unifying and refactoring all kinds of code; I wouldn't be surprised if another refactoring pass would deconstruct and move it again, but I haven't been in that part of the code for a while.)

File: frontend/src/test/java/org/abs_models/frontend/typesystem/LocationTypeTests.java (hunk 5 above)
Comment (rudi) [reply frontend/src/test/java/org/abs_models/frontend/typesystem/LocationTypeTests.java:238]: Ok, thanks.

File: frontend/src/test/java/org/abs_models/frontend/typesystem/LocationTypeTests.java (hunk 6 above)
Comment (rudi) [reply frontend/src/test/java/org/abs_models/frontend/typesystem/LocationTypeTests.java:321]: Ok, thanks.

File: frontend/src/main/java/org/abs_models/frontend/typechecker/locationtypes/LocationTypeExtension.java (hunk 4 above)
Comment (DerDrodt) [reply frontend/src/main/java/org/abs_models/frontend/typechecker/locationtypes/LocationTypeExtension.java:1]: It doesn't really mutate the AST. It simply adds metadata to the `Type` object.

**Comment by rudi** [comment]
Trying to run the tests via `./gradlew build` fails with the following:

```
> Task :frontend:compileTestJava FAILED
/Users/rudi/Source/abstools/frontend/src/test/java/org/abs_models/frontend/typesystem/LocationTypeTests.java:350: error: cannot find symbol
            assertTrue(t.toString(), expected == LocationType.FAR ? t == LocationType.FAR || t.isParametricFar() : expected == t);
                                                                                              ^
  symbol:   method isParametricFar()
  location: variable t of type LocationType
```

**Comment by rudi** [comment]
Many thanks!  I've now merged this pull request.
I'll likely switch on location type checking by default once the unit tests pass -- some old models will certainly fail when we switch on location type inference.

I also wrote a short documentation section in the Types section - any input on how to make this better is appreciated :)

**Comment by rudi** [comment]
I've just filed #290 - @DerDrodt could you have a look?  Thanks in advance!

**Comment by rudi** [comment]
Ok, so I had a closer look on why we have `LocationTypeInferenceExtension` and `LocationTypeExtension`, and I don't really like what I found.

TL;DR: we run the type checker twice.

So we have some analysis stages before type checking; see the reasonably horrible code here: https://github.com/abstools/abstools/blob/d89259841db7109d7ba6dcd654a6c5b1d56b763c/frontend/src/main/java/org/abs_models/frontend/parser/Main.java#L213

Now location type inference is not done in that method; instead `LocationTypeInferenceExtension` walks the tree during type checking.  And since this information is needed for location type checking, `LocationTypeExtension` is not registered as a type checker extension in the normal way; instead, `LocationTypeInferenceExtension` starts a second type-checking pass here: https://github.com/abstools/abstools/blob/d89259841db7109d7ba6dcd654a6c5b1d56b763c/frontend/src/main/java/org/abs_models/frontend/typechecker/locationtypes/LocationTypeInferenceExtension.java#L177

Now we do that only if there are no errors found (or we would double-report errors), and we deregister all other type checker extensions so we don't do so much double work.  Except ... `Model.typeCheck` (https://github.com/abstools/abstools/blob/d89259841db7109d7ba6dcd654a6c5b1d56b763c/frontend/src/main/java/org/abs_models/frontend/typechecker/ext/TypeCheckerExtension.jadd#L348) calls `typeCheckStarted` (https://github.com/abstools/abstools/blob/d89259841db7109d7ba6dcd654a6c5b1d56b763c/frontend/src/main/java/org/abs_models/frontend/typechecker/ext/TypeExtensionHelper.java#L50) which re-registers all those extensions anew.  

So, we have the following bugs:
- If the code contains any other errors, we will not report location type errors because we don't run the checker.
- If we fix this, we will double-report other errors, because ...
- We run all type checking passes two times

TL;DR: The code that `LocationTypeInferenceExtension` runs should be run _before_ type-checking, and `LocationTypeExtension` should be a normal type extension.
